conjure-java-jaxrs-client/src/main/java/com/palantir/conjure/java/client/jaxrs/DialogueFeignClient.java

```diff
@@ -0,0 +1,346 @@
+        if (!maybeContentType.isPresent()) {
+            if (requestBodyContent.length == 0) {
+                return Optional.empty();
+            }
+            throw new SafeIllegalStateException("A Content-Type header was not present on feign request with a body");
+        }
+        return Optional.of(new ByteArrayRequestBody(requestBodyContent, maybeContentType.get()));
+    }
+
+    private static Optional<String> getFirstHeader(Request request, String name) {
+        Collection<String> values = request.headers().get(name);
+        if (values != null) {
+            return Optional.ofNullable(Iterables.getFirst(values, null));
+        }
+        return Optional.empty();
+    }
+
+    private static final class ByteArrayRequestBody implements RequestBody {
+
+        private final byte[] buffer;
+        private final String contentType;
+
+        ByteArrayRequestBody(byte[] buffer, String contentType) {
+            this.buffer = buffer;
+            this.contentType = contentType;
+        }
+
+        @Override
+        public void writeTo(OutputStream output) throws IOException {
+            output.write(buffer);
+        }
+
+        @Override
+        public String contentType() {
+            return contentType;
+        }
+
+        @Override
+        public boolean repeatable() {
+            return true;
+        }
+
+        @Override
+        public void close() {
+            // nothing to do
+        }
+    }
+
+    private static final class DialogueResponseBody implements feign.Response.Body {
+
+        private final Response response;
+
+        DialogueResponseBody(Response response) {
+            this.response = response;
+        }
+
+        @Override
+        public Integer length() {
+            return response.getFirstHeader(HttpHeaders.CONTENT_LENGTH)
+                    .map(Ints::tryParse)
+                    .orElse(null);
+        }
+
+        @Override
+        public boolean isRepeatable() {
+            return false;
+        }
+
+        @Override
+        public InputStream asInputStream() {
+            return response.body();
+        }
+
+        @Override
+        public Reader asReader() {
+            return new InputStreamReader(asInputStream(), StandardCharsets.UTF_8);
+        }
+
+        @Override
+        public void close() {
+            response.close();
+        }
+    }
+
+    enum IdentityDeserializer implements Deserializer<Response> {
+        INSTANCE;
+
+        @Override
+        public Response deserialize(Response response) {
+            return response;
+        }
+
+        @Override
+        public Optional<String> accepts() {
+            return Optional.empty();
+        }
+    }
+
+    /** Implements exception handling equivalent dialogue decoders. */
+    static final class RemoteExceptionDecoder implements feign.codec.ErrorDecoder {
+        static final feign.codec.ErrorDecoder INSTANCE = new RemoteExceptionDecoder();
```

Comment:
any reason not to use the enum pattern here?

Reply:
Good call, originally I had implemented this as a channel that transformed responses into exceptions.